### PR TITLE
Recollect totals on customer address save #10567

### PR DESCRIPTION
**Pull request by tprocter-ldg**
Recollect totals on customer address save. This refreshes the minicart and cart totals when a new customer group is assigned on address save. The customer group can have different tax classes assigned to it and this pull request will recollect the totals to reflect these changes. 

The outer if statement has been removed from the getRateRequest() method so that if the customerId is not set then it retrieves the customer tax class from the customer session. 

An update to the Customer/../sections.xml so that on address formPost, the cart and checkout-totals are refreshed. 

And in the afterSave() method the recollect totals flag is set. 

### Fixed Issues (if relevant)

1. https://github.com/magento/magento2/issues/10567

Testing steps can be found on the above link. (It's quite lengthy sorry!)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=11217) <br/>All committers have signed the CLA.

File: app/code/Magento/Customer/Model/Address.php [hunk 1]
```diff
@@ -329,6 +338,9 @@ public function afterSave()
         if ($indexer->getState()->getStatus() == StateInterface::STATUS_VALID) {
             $this->_getResource()->addCommitCallback([$this, 'reindex']);
         }
+        $quote = $this->checkoutSession->getQuote();
+        $quote->setTriggerRecollect(true);
+        $quote->getResource()->save($quote);
```
Comment (convenient) [inline app/code/Magento/Customer/Model/Address.php:343]: I don't work for magento but have worked on this kind of issue before.

Would it not be safer to apply the `trigger_recollect` flag on the quote, and let the next interaction of the checkout session handled the collection and saving as part of its normal flow?

**Comment by vrann** [comment]
@tprocter-ldg can you please clarify steps to reproduce:

Why should the price be different for the General and Zero VAT Customer groups if 'Rate Percent: 0.00' and no other settings are enabled out of the box?

**Comment by tprocter-ldg** [comment]
Hi @vrann

Steps to reproduce:
Product example being used:
 - Including VAT price 447,00 €
 - Excluding VAT price 369,42 €

1. Go to the frontend as a guest and add an item to your basket.
2. Minicart shows product as 447,00 €
3. Go to login as a new customer.
4. Add your address & vat number when prompted (default functionality).
5. You should receive the 'Your vat id has been successfully validated. You will not be charged tax' message. At this point you are assigned the Zero VAT customer group.
6. Go back to the frontend and notice that all the prices in the category and product pages are now showing the VAT discount (369,42 €).
7. Go to the cart and/or the minicart to view the product you added prior to registration

Expected result:
1. Prices and totals are updated to reflect the VAT discount
ie: product price and total should be 369,42 €.
2. Add another product and view the cart and/or minicart. This product should show the VAT discount price which is reflected on the product page.

Actual result:
1. Prices and totals still show the full price in cart/minicart (447,00 €).
2. Add another product and view the cart and/or minicart. This product also shows as including full vat.
3. Go through to the checkout. Choose your shipping option and click next button.
4. Before proceeding with payment go back to the homepage.
5. Click the minicart and update your product quantity.
6. The prices and totals refresh to show the correct discounted price (369,42 €).

File: app/code/Magento/Customer/Model/Address.php [hunk 2]
```diff
@@ -96,11 +103,13 @@ public function __construct(
         \Magento\Framework\Indexer\IndexerRegistry $indexerRegistry,
         \Magento\Framework\Model\ResourceModel\AbstractResource $resource = null,
         \Magento\Framework\Data\Collection\AbstractDb $resourceCollection = null,
+        Session $checkoutSession,
```
Comment (vrann) [inline app/code/Magento/Customer/Model/Address.php:106]: please introduce dependency in backward compatible way
http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/#adding-a-constructor-parameter

**Review by vrann — CHANGES_REQUESTED** [review]

**Comment by TomashKhamlai** [comment]
Hello, @tprocter-ldg, I found a problem when tried to test this code in real usecases.
Can you try these steps on your developing environment? I've got Fatal Error when tried to save Customer with VAT number on CentOS with
PHP 7.0.23 from the Webtatic repo. Have any idea why this could happen?

**Steps to Reproduce:**

1. Go to Stores -> Taxes -> Tax Rules
2. Click button 'Add New Tax Rule'

> Tax Rule Information:
- Name: Zero VAT
- Tax Rate: Click ‘Add New Tax Rate’
- Tax Identifier: Zero VAT
- Zip/Post is Range: unticked
- Zip/Post Code: *
- State: *
- Country: United States
- Rate Percent: 0.00
> Additional Settings:
- Customer Tax Class: Click ‘Add New Tax Class’
- Zero VAT Customer

Press the tick
Ensure this is the only tax class highlighted

- Product Tax Class: Click ‘Add New Tax
- Zero VAT Product

Press the tick
Ensure this is the only tax class highlighted
- Priority: 0
- Calculate Off Subtotal Only: unticked
- Sort Order: 0

'Save Rule'
3. Go to Stores -> Other Settings -> Customer Groups
4. Click 'Add New Customer Group'
> Group Information:
- Group Name: Zero VAT Customer Group
- Tax Class: Zero VAT Customer

'Save Customer Group'
5. Go to Stores -> Configuration -> Customers -> Customer Configuration -> Create New Account Options
- Enable Automatic Assignment to Customer Group: YES
- Group for Valid VAT ID - Domestic: Zero VAT Customer Group
- Group for Valid VAT ID - Intra-Union: Zero VAT Customer Group
- Group for Invalid VAT ID: General
- Validation Error Group: General
- Validate on Each Transaction: NO
- Show VAT Number on Storefront: YES

'Save Config'
6, Go to Stores -> Configuration -> Sales -> Tax
> Calculation Settings:
- Catalog Prices: (!) Including Tax
- Shipping Prices: (!) Including Tax
- Apply Discount On Prices: (!) Including Tax
> Price Display Settings
- Display Product Prices In Catalog: Including Tax
- Display Shipping Prices: Including Tax
> Shopping Cart Display Settings
- Display Prices: Including and Excluding Tax
- Display Subtotal: Including and Excluding Tax
- Display Shipping Amount : (!) Including Tax

'Save Config'
7. Go to Stores -> Configuration -> General -> Currency Setup -> Currency Options
- Base Currency: Euro
- Default Display Currency: Euro

Make sure that US Dollar and Euro are enabled in Allowed Currencies multiselect list
'Save Config'
8. Go to Catalog -> Catalog -> Products
9. Click 'Add New Product'
- Product Name: simple1
- SKU: simple1
- Price: 447€
- Tax Class: Taxable Goods
- Quantity: 1000
- Weight: 1

Categories: Click 'New Category'
- Category Name: cat1
- Parent Category: Default Category
- 'Create Category'

'Save & Close'
10. Go to Storefront
11. Add simple1 product to the Shopping Cart
12. Click 'Create an Account'
- First Name: John
- Last Name: Doe
- Email: John.Doe@example.com
- Password: Any(Password Strength should be Strong)

'Create an Account'
(!) If you are a registered VAT customer, please click *here* to enter your billing address for proper VAT calculation.
13. Click on 'click here' link and then fill all the necessary fields and VAT Number
- Phone Number: 555-55-555-55
- Street Address: 6161, West Sentinella
- VAT Number: 815302464
- City: Culver City
- State/Province: California
- Zip/Postal Code: 90230
- Country: United States

'Save Address'

Expected result: You saved the address.
Actual result: ???

**Comment by vrann** [comment]
@tprocter-ldg can you please help us to test this issue?

**Comment by okorshenko** [comment]
Hi @tprocter-ldg 
We are closing this PR due to inactivity. Feel free to reopen this PR once ready

**Pull request closed without merging**